dora/core/common/src/main/java/alluxio/uri/UfsUrl.java

```diff
@@ -0,0 +1,486 @@
+        } else {
+          return rootUrl.get("authority");
+        }
+      } else {
+        return inputUrl.get("authority");
+      }
+    }
+  }
+
+  /**
+   * Determines the path components list of this absolute path.
+   * @param rootUrl the elems map of root url, including scheme, authority and path
+   * @param inputUrl the elems map of input url, including scheme, authority and path
+   * @return the path components list of this UfsUrl Object
+   */
+  public static List<String> generatePathComponents(Map<String, String> rootUrl,
+                                                    Map<String, String> inputUrl) {
+    List<String> rootPathComponents = Arrays.asList(rootUrl.get("path").split(PATH_SEPARATOR));
+    List<String> inputPathComponents = Arrays.asList(inputUrl.get("path").split(PATH_SEPARATOR));
+
+    if (rootUrl.equals(inputUrl)) {
+      return rootPathComponents;
+    }
+
+    // inputUrl.scheme is empty -> the inputUrl is a relative path.
+    if (inputUrl.get("scheme").isEmpty()) {
+      rootPathComponents.addAll(inputPathComponents);
+      return rootPathComponents;
+    } else {  // inputUrl is an absolute path.
+      // if inputUrl.path is "/" or "//", i.e., root directory, set a "" to represent root.
+      if (inputUrl.get("path").equals(PATH_SEPARATOR)
+          || inputUrl.get("path").equals(DOUBLE_SLASH_SEPARATOR))  {
+        inputPathComponents = Arrays.asList("");
+      }
+      return inputPathComponents;
+    }
+  }
+
+  /**
+   * Constructs an UfsUrlMessage from a path String.
+   * @param ufsPath a string representing a ufs path
+   * @return an UfsUrlMessage
+   */
+  public static UfsUrlMessage toProto(String ufsPath) {
+    Map<String, String> elements = extractElements(ufsPath);
+
+    Preconditions.checkArgument(
+        elements.containsKey("scheme") && elements.get("scheme") != null
+        && elements.containsKey("authority") && elements.get("authority") != null
+        && elements.containsKey("path") && elements.get("path") != null);
+
+    List<String> pathComponents = Arrays.asList(elements.get("path").split(PATH_SEPARATOR));
+
+    return UfsUrlMessage.newBuilder()
+        .setScheme(elements.get("scheme"))
+        .setAuthority(elements.get("authority"))
+        .addAllPathComponents(pathComponents).build();
+  }
+
+  /**
+   * Creates an UfsUrl instance from a UfsUrlMessage.
+   * @param proto an UfsUrlMessage
+   * @return an UfsUrl object
+   */
+  public static UfsUrl createInstance(UfsUrlMessage proto) {
+    return new UfsUrl(proto);
+  }
+
+  /**
+   * Creates an UfsUrl instance from a String.
+   *
+   * @param ufsPath an input String representing the ufsPath
+   * @return an UfsUrl representing the input String
+   */
+  public static UfsUrl createInstance(String ufsPath) {
+    Preconditions.checkArgument(ufsPath != null && !ufsPath.isEmpty(),
+        "input path is null or empty");
+
+    String ufsRootDir = Configuration.getString(PropertyKey.DORA_CLIENT_UFS_ROOT);
```

Comment:
TL;DR: since this PR is only the first step in a series of related refactoring, let's not rush to providing convenient utility methods and try to look at things as a whole once `AlluxioUrl` and other things are ready.

After some thinking, I think involving `DORA_CLIENT_UFS_ROOT` is not within the scope of this PR.

What `LsCommand` and the like of it need is twofold:

1. parse an arbitrary string provided by the user into a URL or path;
2. if the URL or path is not already a UFS URL, (i.e. a relative path, or an Alluxio URL, or an implicit absolute path within Alluxio namespace), make it a UFS URL by resolving it against a base URL, or even a mount table.

A good URL or path implementation should focus on 1, and make no assumptions about 2. No attempt should be made to try to do 1 and 2 at once in one place. `AlluxioURI` used to do a good job in this regard. 

Today the resolution of a UFS URL requires a root UFS. If tomorrow it requires a mount table, you would soon find yourself trying to inject a mount table into this `createInstance` method. Before long, this method would evolve into a monster that even tries to load configuration from etcd because it needs to retrieve the mount table first.

If not handling the resolution in this class would cause code duplication somewhere else, then let someone else worry about it. Again, I'm not encouraging code duplication. Someone else who authors `LsCommand` and `DoraCacheFileSystem` should come up with appropriate abstractions that allow efficient and clean code reuse.

A possible way out can be: `FileSystemContext` holds the root UFS as a `UfsUrl`, as well as the mount table. `LsCommand` parses user input as a `RelativePath`, and joins the path with the root UFS, or have the mount table turn it into a `UfsUrl`.